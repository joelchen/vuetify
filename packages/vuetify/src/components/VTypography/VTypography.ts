// Types
import Vue, { PropOptions } from 'vue'
import { PropValidator } from 'vue/types/options'

type Scale = 'display-4' | 'display-3' | 'display-2' | 'display-1' | 'headline' | 'title' | 'subtitle-1' | 'subtitle-2' | 'body-1' | 'body-2' | 'caption' | 'overline'
type Breakpoint = 'sm' | 'md' | 'lg' | 'xl'

const GRID_BREAKPOINTS: Breakpoint[] = ['sm', 'md', 'lg', 'xl']

function breakpointProps (): Record<Breakpoint, PropOptions> {
  const options: Dictionary<PropOptions> = {}

  return GRID_BREAKPOINTS.reduce((props, breakpoint) => {
    props[breakpoint] = { type: String }

    return props
  }, options)
}

export default Vue.extend({
  name: 'VTypography',

  functional: true,

  props: {
    ...breakpointProps(),
    scale: {
      type: String,
      default: 'body-1',
    } as PropValidator<Scale>,
    tag: {
      type: String,
      default: 'div',
    },
  },

  render (h, { props, data, children }) {
    const classes: string[] = [props.scale]

    for (const breakpoint of GRID_BREAKPOINTS) {
      const prop = props[breakpoint]

      /* istanbul ignore else */
      if (prop) classes.push(`text-${breakpoint}-${prop}`)
    }

    return h(props.tag, {
      class: classes,
      ...data,
    }, children)
  },
})
