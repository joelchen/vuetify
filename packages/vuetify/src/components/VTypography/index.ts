import VTypography from './VTypography'

export { VTypography }
