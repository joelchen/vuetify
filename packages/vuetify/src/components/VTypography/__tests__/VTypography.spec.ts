// Components
import VTypography from '../VTypography'

// Utilities
import {
  mount,
  MountOptions,
  Wrapper,
} from '@vue/test-utils'

describe('VTypography.ts', () => {
  type Instance = InstanceType<typeof VTypography>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(VTypography, {
        ...options,
      })
    }
  })

  it('should apply correct classes', () => {
    const wrapper = mountFunction({
      propsData: {
        scale: 'headline',
        sm: 'display-1',
        md: 'display-2',
        lg: 'display-3',
        xl: 'display-4',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
