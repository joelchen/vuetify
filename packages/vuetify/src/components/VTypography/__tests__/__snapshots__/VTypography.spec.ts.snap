// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTypography.ts should apply correct classes 1`] = `
<div class="headline text-sm-display-1 text-md-display-2 text-lg-display-3 text-xl-display-4">
</div>
`;
